### PR TITLE
[Post-Release] Move selection of trips to core

### PR DESCRIPTION
**Pull request by bstoeger**
### Describe the pull request:

- [ ] Bug fix
- [ ] Functional change
- [ ] New feature
- [x] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:

In analogy to dives, move the management of trip selection to the core. The main reason is that checking whether a single trip is selected can be painfully slow, which in turn slows down updating of the main tab.

For a 3500 dive test file, this PR reduced the maintab update from 130 to 5 ms, when selecting all dives.

Sadly, the first CTRL-A is still very slow. It appears that the proxy-model has to populate its cache or something. I will have to investigate further.

### Changes made:

1) Introduce a trip->selected flag in analogy to dive->selected.
2) Use said flag to improve speed when many dives are selected.

### Release note:

Probably not.
### Documentation change:

No.

### CODE REVIEW

File: core/selection.cpp [hunk 1]
```diff
@@ -231,3 +237,19 @@ extern "C" void select_newest_visible_dive()
 	// No visible dive -> deselect all
 	select_single_dive(nullptr);
 }
+
+extern "C" void select_trip(struct dive_trip *trip)
```
Comment (dirkhh) [inline core/selection.cpp:241]: so in this commit I only see a call to `deselect_trip()` and no calls to `select_trip()` which makes me wonder... but most likely that's simply in the next commits?

File: desktop-widgets/divelistview.cpp [hunk 2]
```diff
@@ -566,9 +569,11 @@ void DiveListView::selectionChanged(const QItemSelection &selected, const QItemS
 		const QAbstractItemModel *model = index.model();
 		struct dive *dive = model->data(index, DiveTripModelBase::DIVE_ROLE).value<struct dive *>();
 		if (!dive) { // it's a trip!
+			dive_trip *trip = model->data(index, DiveTripModelBase::TRIP_ROLE).value<dive_trip *>();
+			select_trip(trip);
```
Comment (dirkhh) [inline desktop-widgets/divelistview.cpp:555]: ahhh, here it is

File: core/selection.cpp [hunk 3]
```diff
@@ -254,6 +254,17 @@ extern "C" void deselect_trip(struct dive_trip *trip)
...
 
+extern "C" struct dive_trip *single_selected_trip()
+{
+	if (amount_trips_selected != 1)
+		return NULL;
+	for (int i = 0; i < trip_table.nr; ++i) {
+		if (trip_table.trips[i]->selected)
+			return trip_table.trips[i];
+	}
+	return NULL; // shouldn't happen
```
Comment (dirkhh) [inline core/selection.cpp:266]: print out a warning?

File: core/selection.cpp [hunk 4]
```diff
@@ -231,3 +237,19 @@ extern "C" void select_newest_visible_dive()
 	// No visible dive -> deselect all
 	select_single_dive(nullptr);
 }
+
+extern "C" void select_trip(struct dive_trip *trip)
+{
+	if (trip && !trip->selected) {
+		trip->selected = true;
+		amount_trips_selected++;
```
Comment (dirkhh) [inline core/selection.cpp:245]: in our UI logic, selecting a trip always implies selecting all the dives in the trip.
Would it make sense to do this hear, or would this confuse other layers and we really just worry about the helper data structure for trips?
As I write the question, I think I know the answer... :-)

**Review by dirkhh — APPROVED** [review]
I think this looks good.
I'm happy to see you work on all these post-release things already...
I cannot wait to get this release out, deal with the almost guaranteed random breakage that we didn't catch (because... 1500+ commits), and move forward with cool new stuff.

File: core/selection.cpp (hunk 4 above)
Comment (bstoeger) [reply core/selection.cpp:245]: Yes, that was the original idea. However, I'm not sure about the correct way to implement this, so first I wanted to start with the "easy" version. The problem is that we get a loop:

UI selection change -> dive list -> Core selects more items -> dive list

We end up with a selection change in the divelist, while a selection change is active. What we could do is
1) a "apply_selection_change()" function that returns a list of additional items that the dive list has to select.
2) simply intercepting all UI selection changes, sending them to the core and let the core rebuild the selection. I.e. never apply selections directly.

Comment (dirkhh) [reply core/selection.cpp:245]: I can see the advantage of the much simpler approach to not deal with causing new selections

**Comment by dirkhh** [comment]
OK, this is good to go - unless you want to add the warning that I suggested.
And I'll wait for #2758

File: core/selection.cpp (hunk 3 above)
Comment (bstoeger) [reply core/selection.cpp:266]: done.

Comment (dirkhh) [reply core/selection.cpp:266]: it must be my day to be blind... I don't see it...

Comment (bstoeger) [reply core/selection.cpp:266]: Forgot "-a" in "commit --amend".

**Comment by dirkhh** [comment]
This is so frustrating. TestGitStorage failed. But only in the mobile build on Ubuntu.
No useful errors. Works in the other builds. And anyway, this shouldn't affect anything in the storage test.
I try to rerun the test, and get the (%*&@(#% this isn't a tree error again. What the heck.

**Pull request merged**